js-peer/src/lib/direct-message.ts

```diff
@@ -0,0 +1,180 @@
+  registrar: Registrar
+  connectionManager: ConnectionManager
+}
+
+export class DirectMessage extends TypedEventEmitter<DirectMessageEvents> implements Startable {
+  readonly [serviceDependencies]: string[] = [
+    '@libp2p/identify',
+    '@libp2p/connection-encryption',
+    '@libp2p/transport',
+    '@libp2p/stream-multiplexing',
+  ]
+
+  readonly [serviceCapabilities]: string[] = [
+    '@universal-connectivity/direct-message'
+  ]
+
+  private topologyId?: string
+  private readonly components: DirectMessageComponents
+  private dmPeers: Set<string> = new Set()
+
+  constructor(components: DirectMessageComponents) {
+    super()
+    this.components = components
+  }
+
+  async start(): Promise<void> {
+    this.topologyId = await this.components.registrar.register(DIRECT_MESSAGE_PROTOCOL, {
+      onConnect: this.handleConnect.bind(this),
+      onDisconnect: this.handleDisconnect.bind(this)
+    })
+  }
+
+  async afterStart(): Promise<void> {
+    await this.components.registrar.handle(
+      DIRECT_MESSAGE_PROTOCOL,
+      async ({ stream, connection }) => {
+        await this.receive(stream, connection)
+      },
+    )
+  }
+
+  stop(): void {
+    if (this.topologyId != null) {
+      this.components.registrar.unregister(this.topologyId)
+    }
+  }
+
+  private handleConnect(peerId: PeerId): void {
+    this.dmPeers.add(peerId.toString())
+  }
+
+  private handleDisconnect(peerId: PeerId): void {
+    this.dmPeers.delete(peerId.toString())
+  }
+
+  isDMPeer(peerId: PeerId): boolean {
+    return this.dmPeers.has(peerId.toString())
+  }
+
+  async send(peerId: PeerId, message: string): Promise<boolean> {
+    if (!message) {
+      throw new Error(ERRORS.EMPTY_MESSAGE)
+    }
+
+    let stream: Stream | undefined
+
+    try {
+      const conn = await this.components.connectionManager.openConnection(peerId, { signal: AbortSignal.timeout(5000) })
+      if (!conn) {
+        throw new Error(ERRORS.NO_CONNECTION)
+      }
+
+      // Single protocols can skip full negotiation
+      const stream = await conn.newStream(DIRECT_MESSAGE_PROTOCOL, { negotiateFully: false })
+
+      if (!stream) {
+        throw new Error(ERRORS.NO_STREAM)
+      }
+
+      const datastream = pbStream(stream)
+
+      const req: dm.DirectMessageRequest = {
+        content: message,
+        type: MIME_TEXT_PLAIN,
+        metadata: {
+          clientVersion: dmClientVersion,
+          timestamp: BigInt(Date.now()),
+        },
+      }
+
+      await datastream.write(req, dm.DirectMessageRequest)
+
+      const res = await datastream.read(dm.DirectMessageResponse)
+
+      if (!res) {
+        throw new Error(ERRORS.NO_RESPONSE)
+      }
+
+      if (!res.metadata) {
+        throw new Error(ERRORS.NO_METADATA)
+      }
+
+      if (res.status !== dm.Status.OK) {
+        throw new Error(ERRORS.STATUS_NOT_OK(res.status))
+      }
+    } catch (e: any) {
+      stream?.abort(e)
+      throw e
+    } finally {
+      await stream?.close()
```

Comment:
I'm curious what the tradeoffs of opening and closing a stream for every message are. Not a blocker, but useful for learning.

Reply:
You’re incurring a small amount of overhead - there will be a round trip for opening a stream, two for multistream select (e.g. negotiating a protocol) and then one for closing the stream afterwards.

This may not be disastrous depending on the frequency of message sending and your network latency to the remote peer, it’s application-specific.

Reply:
Another factor is that opening/closing streams can be useful if the remote has a strict limit on the number of incoming protocol streams.

Reply:
> You’re incurring a small amount of overhead - there will be a round trip for opening a stream, two for multistream select (e.g. negotiating a protocol) and then one for closing the stream afterwards.

If the connection is open already, and `const stream = await conn.newStream(DIRECT_MESSAGE_PROTOCOL, { negotiateFully: false })` is called (line 106), will it entail two roundtrips for multistream select?

Reply:
Ah yeah - `negotiateFully: false` will just send the mss header, the line break, and the protocol name in one go, assume success, and carry on so yeah, it doesn't wait to read anything.  It'll still wait for the message to be passed to the network layer but that's pretty fast.

Reply:
My thought on this was that opening/closing a stream would have a relatively low overhead compared to opening a connection - it also simplifies the code a bit by having one less thing to track / cleanup.

I'm not sure how the connection manager handles connections but would assume that it wouldn't close connections with an open stream, thus best to close the stream when possible.

Reply:
> would assume that it wouldn't close connections with an open stream

I believe that's correct

>  thus best to close the stream when possible.

mm, not sure I understood what you mean. Do you mean that you rely on another open stream, i.e.the gossipsub stream that remains open to the peer for the public chat?

Reply:
> > would assume that it wouldn't close connections with an open stream
> 
> I believe that's correct
> 
> > thus best to close the stream when possible.
> 
> mm, not sure I understood what you mean. Do you mean that you rely on another open stream, i.e.the gossipsub stream that remains open to the peer for the public chat?

I mean so that it frees the stream resource, so that the connection manager can choose to close the DM connection if needed.